lms/djangoapps/commerce/signals.py

```diff
@@ -0,0 +1,107 @@
+"""
+Signal handling functions for use with external commerce service.
+"""
+import logging
+
+from django.dispatch import receiver
+from ecommerce_api_client.exceptions import HttpClientError
+from request_cache.middleware import RequestCache
+from student.models import UNENROLL_DONE
+
+from commerce import ecommerce_api_client, is_commerce_service_configured
+
+log = logging.getLogger(__name__)
+
+
+@receiver(UNENROLL_DONE)
+def handle_unenroll_done(sender, course_enrollment=None, skip_refund=False, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal receiver for unenrollments, used to automatically initiate refunds
+    when applicable.
+
+    N.B. this signal is also consumed by lms.djangoapps.shoppingcart.
+    """
+    if not is_commerce_service_configured():
+        return
+
+    if course_enrollment and course_enrollment.refundable() and not skip_refund:
+        try:
+            request_user = get_request_user() or course_enrollment.user
+            refund_seat(course_enrollment, request_user)
+        except:  # pylint: disable=bare-except
+            # don't assume the signal was fired with `send_robust`.
+            # avoid blowing up other signal handlers by gracefully
+            # trapping the Exception and logging an error.
+            log.exception(
+                "Unexpected exception while attempting to initiate refund for user [%s], course [%s]",
+                course_enrollment.user,
+                course_enrollment.course_id,
+            )
+
+
+def get_request_user():
+    """
+    Helper to get the authenticated user from the current HTTP request (if
+    applicable).
+
+    If the requester of an unenrollment is not the same person as the student
+    being unenrolled, we authenticate to the commerce service as the requester.
+    """
+    request = getattr(RequestCache.get_request_cache(), 'request')
```

Comment:
Can I suggest that you simply make this a classmethod on RequestCache? I think exposing the internals of RequestCache is a bit messy here. How about, in request_cache:

@classmethod
def get_current_thread_request()

@classmethod
def get_current_thread_user()